### PR TITLE
Fixes #28868 - Add location inherited parameters

### PR DESCRIPTION
**Pull request by shiramax**
Adding location inherited parameters to a GET request.

### CODE REVIEW

**Comment by shiramax** [comment]
[test katello]

**Review by ofedoren — COMMENTED** [review]
Works well with hammer (https://github.com/theforeman/hammer-cli-foreman/pull/496). But I'm not sure if Katello test failures are related..

**Comment by timogoebel** [comment]
Is there a noticeable performance penalty with this?

**Comment by shiramax** [comment]
https://github.com/Katello/katello/pull/8568 Should Fix the failing tests.

**Comment by ezr-ondrej** [comment]
> Is there a noticeable performance penalty with this?

I wonder the same, though this is probably not piped to other commands, only used by user if he wants to see information about the object, so I would say it doesn't matter if it runs bit longer for complete information. As non of the added relations shouldn't hit `> 100` of records, I would say the time should be reasonable.

File: app/views/api/v2/taxonomies/show.json.rabl [hunk 1]
```diff
@@ -6,6 +6,7 @@ attribute :ignore_types => :select_all_types
 
 attributes :description, :created_at, :updated_at
 
+ancestors = @taxonomy.ancestors
```
Comment (ezr-ondrej) [inline app/views/api/v2/taxonomies/show.json.rabl]:
```suggestion
ancestors = @taxonomy.ancestors.preload(:users, :smart_proxies, :subnets, :compute_resources ....)
```

**Review by ezr-ondrej — COMMENTED** [review]
Would it make sense to preload the relations?

File: app/views/api/v2/taxonomies/show.json.rabl (hunk 1 above)
Comment (shiramax) [reply app/views/api/v2/taxonomies/show.json.rabl]: @ezr-ondrej done:)

**Review by ezr-ondrej — DISMISSED** [review]
I think the benefit outweights the performance here. Thanks @shiramax, :+1: from me.

**Comment by ezr-ondrej** [comment]
Last idea (feel free to disagree): would it make sense to join the inherited resources and only add an attribute "inherited=true" that would make sense to me as an API user as I am interested in what resources are connected with this taxonomy, I do not care about if they are inherited or not that much :thinking:

**Comment by shiramax** [comment]
@ezr-ondrej 
mm, I can see that it solves the performance issue, however, I feel it's not the correct approach here, 
I believe that when A user wants to get the information on a specific location, then he should get all the inherited attributes as well.

**Comment by ezr-ondrej** [comment]
I meant if we could rendered it in the same group, just having a flag, that would tag them as inherited.

**Comment by shiramax** [comment]
@ezr-ondrej but the users will have to add this flag in order to get the inherited attributes. 
and I think the default beaviour should be to get all the attributes, without adding this flag.

**Comment by ezr-ondrej** [comment]
@shiramax no, there would be no input attribute to filter them (or maybe there could, but default on). But I meant join the output groups as 
```json
{ "users": [{"id": 1, "inherited": false}, {"id": 2, "inherited": true}] },
...
````

**Comment by shiramax** [comment]
@ezr-ondrej, I've changed it so all parameters, inherited or not will be presented under the same name, let me know what you think about this change

**Comment by ezr-ondrej** [comment]
Sorry to be a nosy parker here @shiramax, but is the user able to say now what is inherited resource and what is direct associated resource?

**Comment by shiramax** [comment]
@ezr-ondrej , 
I could add an inherited node as you mentioned :
               [0] {
                             "id" => 5,
                          "login" => "mshira",
                    "description" => "sss",
                      "inherited" => false
                },
                [1] {
                             "id" => 6,
                          "login" => "shira",
                    "description" => "",
                      "inherited" => false
                }
            ]

but when the user will want to perform PUT or POST, he will might be confused and try to send the inherited as well. 

I could do, something like this: 
  "users" => [
        [0] {
               "native" => [
                [0] {
                             "id" => 5,
                          "login" => "mwdshira",
                    "description" => "sss"
                },
                [1] {
                             "id" => 6,
                          "login" => "dshira",
                    "description" => ""
                }
            ],
            "inherited" => [
                [0] {
                             "id" => 9,
                          "login" => "mshira",
                    "description" => "sss"
                },
                [1] {
                             "id" => 10,
                          "login" => "shira",
                    "description" => ""
                }
            ]
        },

but this means I'm changing the output of the API. 

I'm not sure what's the best solution here, and if it is really important for the user to know which parameters are inherited and which are not.

**Comment by ezr-ondrej** [comment]
> I'm not sure what's the best solution here, and if it is really important for the user to know which parameters are inherited and which are not.

It's probably not mission critical, to see it, for that I like the first solution. I got the concern about the confusing user, but we can add just simple `Only direct, omit inherited` to documentation of the POST / PUT requests and I believe it's really straight forward, if you stop and think about it while using the endpoint.

The second one is worse than the original solution IMHO.
For me the main usecase of this endpoint is "Let see what entities are associated with this loc/org" and for that the parameter of the entity seems the best to me, but I agree it's not perfect eighter and quite frankly I'm open to get back to the original solution if you feel it's more convenient.

**Comment by shiramax** [comment]
[test foreman]

**Comment by shiramax** [comment]
@ezr-ondrej done :)

File: app/views/api/v2/taxonomies/show.json.rabl [hunk 2]
```diff
@@ -6,52 +6,83 @@ attribute :ignore_types => :select_all_types
...
+)
 node :hosts_count do |taxonomy|
   hosts_count[taxonomy]
 end
 
-child :users do
+child (@taxonomy.users + ancestors.map(&:users).flatten).uniq => :users do
   extends "api/v2/users/base"
+  node :inherited do |current_user|
+    !@taxonomy.users.map { |user| user.id }.include?(current_user.id)
```
Comment (ezr-ondrej) [inline app/views/api/v2/taxonomies/show.json.rabl]: Small nit: ActiveRecord implements `==` operator to basically `right.insatance_of?(self.class) && left.id == right.id` and we do not customize it anywhere I believe, we could thus reduce this with none performance impact to more readable
```suggestion
    !@taxonomy.users.include?(user)
```

**Review by ezr-ondrej — DISMISSED** [review]
Small nit, but I'm good even with it :+1: 
But now I see I've totally stole it from @tbrisker, I thought it's blocked on me :rofl:
@tbrisker care to take a look and maybe push the button?

File: app/views/api/v2/taxonomies/show.json.rabl (hunk 2 above)
Comment (shiramax) [reply app/views/api/v2/taxonomies/show.json.rabl]: @ezr-ondrej  done :)

**Review by ezr-ondrej — APPROVED** [review]
Works for me :+1:

**Pull request merged**